src/util.c

```diff
@@ -552,6 +552,30 @@ int string2d(const char *s, size_t slen, double *dp) {
     return 1;
 }
 
+/* Returns 1 if the double value can safely be represented in long long without
+ * precision loss, in which case the corresponding long long is stored in the out variable. */
+int double2ll(double d, long long *out) {
+#if (DBL_MANT_DIG >= 52) && (LLONG_MAX == 0x7fffffffffffffffLL)
+        /* Check if the float is in a safe range to be casted into a
+         * long long. We are assuming that long long is 64 bit here.
+         * Also we are assuming that there are no implementations around where
+         * double has precision < 52 bit.
+         *
+         * Under this assumptions we test if a double is inside an interval
+         * where casting to long long is safe. Then using two castings we
+         * make sure the decimal part is zero. If all this is true we can use
+         * integer without precision loss. */
+        double min = -4503599627370495; /* -(2^52)+1 */
+        double max = 4503599627370495; /* (2^52)-1 */
+        if (d < min || d > max)
+            return 0;
+        *out = d;
+        if (*out == d)
+            return 1;
```

Comment:
Most likely it's better to avoid the mem write for fractional numbers, I'd just change it without benchmarking.

```suggestion
        long long ll = d;
        if (ll == d) {
            *out = ll;
            return 1;
        }
```

Reply:
fine... let's not waste too much effort to prove if this actually matters.